Gitweb:     http://git.kernel.org/git/?p=linux/kernel/git/torvalds/linux-2.6.git;a=commit;h=38332cb98772f5ea757e6486bed7ed0381cb5f98
Commit:     38332cb98772f5ea757e6486bed7ed0381cb5f98
Parent:     e48af19f56eb47a1f908ee8f16df9d246f955b21
Author:     Segher Boessenkool <segher@kernel.crashing.org>
AuthorDate: Tue Mar 4 14:59:54 2008 -0800
Committer:  Thomas Gleixner <tglx@linutronix.de>
CommitDate: Sun Mar 9 08:42:57 2008 +0100

    time: prevent the loop in timespec_add_ns() from being optimised away
    
    Since some architectures don't support __udivdi3().
    
    Signed-off-by: Segher Boessenkool <segher@kernel.crashing.org>
    Cc: john stultz <johnstul@us.ibm.com>
    Cc: Ingo Molnar <mingo@elte.hu>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Thomas Gleixner <tglx@linutronix.de>

 include/linux/time.h |    4 ++++
 1 files changed, 4 insertions(+), 0 deletions(-)

diff --git a/include/linux/time.h b/include/linux/time.h
index 2091a19..d32ef0a 100644
--- a/include/linux/time.h
+++ b/include/linux/time.h
@@ -174,6 +174,10 @@ static inline void timespec_add_ns(struct timespec *a, u64 ns)
 {
 	ns += a->tv_nsec;
 	while(unlikely(ns >= NSEC_PER_SEC)) {
+		/* The following asm() prevents the compiler from
+		 * optimising this loop into a modulo operation.  */
+		asm("" : "+r"(ns));
+
 		ns -= NSEC_PER_SEC;
 		a->tv_sec++;
 	}
